### ISSUE TITLE
[v2 QUESTION]: MCC calculation is different to the conventional formula

### ISSUE THREAD

**Issue by kennylam8** [chemprop/chemprop#1294]
Hi, I have a question about the MCC loss / metric calculation in ChemProp, I am currently using the latest version.

In the formula for MCC metric/loss:
```
def _calc_unreduced_loss(self, preds, targets, mask, weights, *args) -> Tensor:
  TP = (targets * preds * weights * mask).sum(0, keepdim=True)
  FP = ((1 - targets) * preds * weights * mask).sum(0, keepdim=True)
  TN = ((1 - targets) * (1 - preds) * weights * mask).sum(0, keepdim=True)
  FN = (targets * (1 - preds) * weights * mask).sum(0, keepdim=True)
  
  return TP, FP, TN, FN
```
For example, in the TP part, does `targets * preds` mean it is taking the predicted probability and multiplying it with the target label? I understand this might be a better metric for validation and testing, but this seems to be off from the conventional form of MCC metric, where the preds is usually the final class prediction but not the probability, maybe it would be better if the metric could be labelled as "soft MCC" or something indicative of the difference. Please correct me if I'm mistaken, thanks!

And thank you for building this amazing tool for the cheminformatics community.

Thanks,
Kenny

**Comment by kennylam8** [comment]
> Without looking at the implementation details, we can check the values against the implement in `sklearn`:
> 
> import numpy as np
> import torch
> 
> mcc_loss = BinaryMCCLoss()
> 
> for i in range(10):
>   np.random.seed(i)
>   true_labels = np.random.randint(0, 2, 128)
>   predictions = np.random.randint(0, 2, 128)
> 
>   matthews_corrcoef_value = matthews_corrcoef(true_labels, predictions)
> 
>   print(f"Matthews Correlation Coefficient for seed {i}: {matthews_corrcoef_value}")
> 
>   true_labels_tensor = torch.tensor(true_labels, dtype=torch.float32)
>   predictions_tensor = torch.tensor(predictions, dtype=torch.float32)
> 
>   chemprop_mcc_value = 1 - mcc_loss(predictions_tensor, true_labels_tensor)
> 
>   print(f"Chemprop MCC value: {chemprop_mcc_value.item()}")
> 
>   deviation_percentage = abs(matthews_corrcoef_value - chemprop_mcc_value.item()) / matthews_corrcoef_value
>   print(f"Deviation percentage {deviation_percentage:%}")
>   print()
> 
> >>> Matthews Correlation Coefficient for seed 0: 0.023121408441593813
> Chemprop MCC value: 0.023121416568756104
> Deviation percentage 0.000035%
> 
> Matthews Correlation Coefficient for seed 1: 0.015719307070490657
> Chemprop MCC value: 0.015719294548034668
> Deviation percentage 0.000080%
> 
> Matthews Correlation Coefficient for seed 2: 0.09263301734563939
> Chemprop MCC value: 0.09263300895690918
> Deviation percentage 0.000009%
> 
> Matthews Correlation Coefficient for seed 3: -0.10024600283175306
> Chemprop MCC value: -0.10024595260620117
> Deviation percentage -0.000050%
> 
> Matthews Correlation Coefficient for seed 4: -0.04739221033946729
> Chemprop MCC value: -0.04739224910736084
> Deviation percentage -0.000082%
> 
> Matthews Correlation Coefficient for seed 5: -0.04688072309384954
> Chemprop MCC value: -0.04688072204589844
> Deviation percentage -0.000002%
> 
> Matthews Correlation Coefficient for seed 6: 0.07821097239258434
> Chemprop MCC value: 0.07821094989776611
> Deviation percentage 0.000029%
> 
> Matthews Correlation Coefficient for seed 7: 0.07291705677762582
> Chemprop MCC value: 0.07291704416275024
> Deviation percentage 0.000017%
> 
> Matthews Correlation Coefficient for seed 8: -0.10731384280044624
> Chemprop MCC value: -0.10731387138366699
> Deviation percentage -0.000027%
> 
> Matthews Correlation Coefficient for seed 9: 0.018892010195205314
> Chemprop MCC value: 0.018891990184783936
> Deviation percentage 0.000106%
> So this implementation is nearly identical to the reference implementation.
> 
> You can see this notebook here to generate these results: https://colab.research.google.com/drive/1rWyxfQ6AZTLiDSLohrRdg70JnSOdX8A2?usp=sharing

Thank you for your reply and help, yes indeed the MCC is identical when we consider integer values, however, in the case of binary classification, the chemprop version is taking in the prediction probability (float) of class 1 as this is suitable for "soft MCC", that is significantly different than sklearn's version which requires integers (final prediction class) for TP/TN/FN/FP.


```
import numpy as np
import torch

mcc_loss = BinaryMCCLoss()

for i in range(10):
  np.random.seed(i)
  true_labels = np.random.randint(0, 2, 1000)
  predictions_proba = np.random.rand(1000) # for probability outputs
  predictions = [1 if x > 0.5 else 0 for x in predictions_proba] # assume we use 0.5 as the class separation cutoff

  matthews_corrcoef_value = matthews_corrcoef(true_labels, predictions) # sklearn only takes integer final class predictions
  

  print(f"Matthews Correlation Coefficient for seed {i}: {matthews_corrcoef_value}")

  true_labels_tensor = torch.tensor(true_labels, dtype=torch.float32)
  predictions_tensor = torch.tensor(predictions_proba, dtype=torch.float32)

  chemprop_mcc_value = 1 - mcc_loss(predictions_tensor, true_labels_tensor) # chemprop takes probability as output
  print(f"Chemprop MCC value: {chemprop_mcc_value.item()}")

  deviation_percentage = abs(matthews_corrcoef_value - chemprop_mcc_value.item()) / matthews_corrcoef_value
  print(f"Deviation percentage {deviation_percentage:%}")
  print()

>>> Matthews Correlation Coefficient for seed 0: -0.02793778801843318
Chemprop MCC value: -0.018834352493286133
Deviation percentage -32.584668%

Matthews Correlation Coefficient for seed 1: 0.009569746037903861
Chemprop MCC value: 0.004167139530181885
Deviation percentage 56.455067%

Matthews Correlation Coefficient for seed 2: 0.013501870292433459
Chemprop MCC value: 0.009920656681060791
Deviation percentage 26.523834%

Matthews Correlation Coefficient for seed 3: 0.0133646388680892
Chemprop MCC value: 0.007615029811859131
Deviation percentage 43.021058%

Matthews Correlation Coefficient for seed 4: -0.008161893674539867
Chemprop MCC value: -0.006451725959777832
Deviation percentage -20.953075%

Matthews Correlation Coefficient for seed 5: -0.008721829756251331
Chemprop MCC value: -0.011810064315795898
Deviation percentage -35.408104%

Matthews Correlation Coefficient for seed 6: -0.005841039722842272
Chemprop MCC value: -0.00910484790802002
Deviation percentage -55.877178%

Matthews Correlation Coefficient for seed 7: 0.027864123922444743
Chemprop MCC value: 0.010102152824401855
Deviation percentage 63.744947%

Matthews Correlation Coefficient for seed 8: -0.005025165911765817
Chemprop MCC value: 0.0017551183700561523
Deviation percentage -134.926576%

Matthews Correlation Coefficient for seed 9: -0.029622516313847342
Chemprop MCC value: -0.021054744720458984
Deviation percentage -28.923173%
```
